src/routes/puzzles/[id]/play/+page.svelte

```diff
@@ -174,43 +176,61 @@
 		<Button on:click={submitCode}>Submit code</Button>
 	</div>
 
+	{#if puzzle.validators}
+		<TestProgressBar validators={puzzle.validators} class="my-4" />
+	{/if}
+
 	{#if puzzle.validators}
 		<Accordion.Root bind:value={testsAccordion}>
 			<Accordion.Item value="tests">
 				<Accordion.Trigger><h2>Tests</h2></Accordion.Trigger>
 				<Accordion.Content>
-					<ul class="flex flex-col gap-8">
+					<ul class="flex flex-col gap-10">
 						{#each puzzle.validators as validator, index}
-							<li
-								class={cn(calculatePuzzleResultColor(validator.testResult?.run.result), "w-full")}
-							>
-								<div class="lg:flex">
-									<div class="lg:w-1/2">
-										<h3 class="font-semibold">Input</h3>
-										<pre>
+							<li class="relative">
+								<div
+									class={cn(
+										calculatePuzzleResultColor(validator.testResult?.run.result),
+										"w-full space-y-8 rounded-lg border-2 p-4"
+									)}
+									id={`validator-${index}`}
+								>
+									<div class="lg:flex">
+										<div class="space-y-2 lg:w-1/2">
+											<h3 class="text-lg font-semibold">Input</h3>
+											<pre>
 											{validator.input.trim()}
 										</pre>
-									</div>
-									<div class="lg:w-1/2">
-										<h3 class="font-semibold">Output</h3>
-										<pre>
+										</div>
+										<div class="space-y-2 lg:w-1/2">
+											<h3 class="text-lg font-semibold">Expected output</h3>
+											<pre>
 											{validator.output}
 										</pre>
+										</div>
 									</div>
-								</div>
 
-								<Button on:click={() => executeCode(index, validator.input, validator.output)}>
-									Run code
-								</Button>
+									<div class="space-y-2">
+										{#if validator.testResult}
+											<h3 class="text-lg font-semibold">Latest result</h3>
+											<div class="lg:w-1/2">
+												<h4 class="font-bold">Stdout:</h4>
+												<pre>{validator.testResult?.run.stdout}</pre>
+											</div>
+											<div class="lg:w-1/2">
+												<h4 class="font-bold">Stderr:</h4>
+												<pre>{validator.testResult?.run.stderr}</pre>
+											</div>
+										{/if}
+									</div>
```

Comment:
It would save some space to omit this entire element when there are no `testResult`s yet

```suggestion
									{#if validator.testResult}
										<div class="space-y-2">
											<h3 class="text-lg font-semibold">Latest result</h3>
											<div class="lg:w-1/2">
												<h4 class="font-bold">Stdout:</h4>
												<pre>{validator.testResult?.run.stdout}</pre>
											</div>
											<div class="lg:w-1/2">
												<h4 class="font-bold">Stderr:</h4>
												<pre>{validator.testResult?.run.stderr}</pre>
											</div>
										</div>
									{/if}
```